### PR TITLE
Add Depth Anything

### PR DESCRIPTION
**Pull request by NielsRogge**
# What does this PR do?

This PR adds an alternative design to #28643, which adds a standalone separate model.

Pros:

- [x] does not clutter the existing modeling_dpt.py
- [x] is more in line with the [philosophy](https://huggingface.co/blog/transformers-design-philosophy)

Cons:

- [x] actually, not a lot :)

### CODE REVIEW

File: src/transformers/models/depth_anything/modeling_depth_anything.py [hunk 1]
```diff
@@ -0,0 +1,471 @@
...
+        # first layer only uses the last hidden_state
+        size = hidden_states[1].shape[2:]
+        fused_hidden_state = self.layers[0](hidden_states[0], size=size)
+        fused_hidden_states.append(fused_hidden_state)
+        # looping from the last layer to the second
+        for idx, (hidden_state, layer) in enumerate(zip(hidden_states[1:], self.layers[1:])):
+            if idx != len(hidden_states[1:]) - 1:
+                size = hidden_states[1:][idx + 1].shape[2:]
+            fused_hidden_state = layer(fused_hidden_state, hidden_state, size=size)
+            fused_hidden_states.append(fused_hidden_state)
```
Comment (amyeroberts) [inline src/transformers/models/depth_anything/modeling_depth_anything.py:142]: This can be simplified without having to use the conditional indexing logic 

```suggestion
        # We go backwards from the last hidden state to the first
        hidden_states = hidden_states[::-1]

        output_sizes = [hidden_state.shape[2:] for hidden_state in hidden_states[1:]]
        output_sizes += output_sizes[-1]

        # first layer only uses the last hidden_state
        residuals = [None] + hidden_states[1:]

        fused_hidden_states = []
        fused_hidden_state = hidden_states[0]
        
        # looping from the second last layer to the first
        for (residual, layer, output_size) in zip(residuals, self.layers, output_sizes):
            fused_hidden_state = layer(fused_hidden_state, residual=residual, size=output_size)
            fused_hidden_states.append(fused_hidden_state)
```

File: src/transformers/models/depth_anything/modeling_depth_anything.py [hunk 2]
```diff
@@ -0,0 +1,471 @@
...
+
+    def __init__(self, config):
+        super().__init__()
+
+        self.config = config
+        self.layers = nn.ModuleList()
+        for i, factor in zip(range(len(config.neck_hidden_sizes)), config.reassemble_factors):
+            self.layers.append(
+                DepthAnythingReassembleLayer(config, channels=config.neck_hidden_sizes[i], factor=factor)
+            )
```
Comment (amyeroberts) [inline src/transformers/models/depth_anything/modeling_depth_anything.py:75]: We can just iterate over `config.neck_hidden_sizes` directly instead of indexing 

```suggestion
        for channels, factor in zip(config.neck_hidden_sizes, config.reassemble_factors):
            self.layers.append(
                DepthAnythingReassembleLayer(config, channels=channels, factor=factor)
            )
```

**Review by amyeroberts — COMMENTED** [review]
Thanks for adding! 

As a separate model is a lot cleaner ❤️ Mostly nits. Main comment is about not using the `backbone_config` to set up the model.

File: src/transformers/models/depth_anything/modeling_depth_anything.py (hunk 1 above)
Comment (NielsRogge) [reply src/transformers/models/depth_anything/modeling_depth_anything.py]: This is not giving equivalent logits :/ there might be a bug in there

**Comment by NielsRogge** [comment]
@amyeroberts thanks for your review, addressed all comments. CI is failing due to unrelated tests

File: tests/models/depth_anything/test_modeling_depth_anything.py [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+    image = Image.open("./tests/fixtures/tests_samples/COCO/000000039769.png")
+    return image
+
+
+@require_torch
+@require_vision
+@slow
+class DepthAnythingModelIntegrationTest(unittest.TestCase):
+    def test_inference(self):
+        # TODO update organization
```
Comment (amyeroberts) [inline tests/models/depth_anything/test_modeling_depth_anything.py:224]: either still to-do or comment should be removed 

```suggestion
```

**Review by amyeroberts — APPROVED** [review]
Thanks for iterating and adding this model!
